climada/hazard/base.py

```diff
@@ -1396,47 +1394,99 @@ def _read_att_excel(self, file_name, var_names):
 
-        It is recommended to use the static method `hazard.concatenate_hazard`
-        which concatenates a list of hazards into a new object.
-
         Parameters
         ----------
-        hazard: climada.hazard.Hazard object
-            Hazard instance to append to self
+        haz_list : one or more climada.hazard.Hazard objects
+            Hazard instances to append to self
 
         Raises
         ------
         TypeError
 
         See Also
         --------
-        hazard.concat: concatenate 2 or more hazards
+        Hazard.concat: concatenate 2 or more hazards
         """
-        hazard._check_events()
-        if type(self) != type(hazard):
-            raise TypeError(f"hazard is of class {type(self)} which is "
-                            f" different from {type(hazard)}")
+        if len(haz_list) == 0:
+            return
+        haz_list = (self,) + haz_list
+        haz_list_nonempty = [haz for haz in haz_list if haz.size > 0]
 
-        if len(self.event_id) == 0: #if self is empty, replace self with hazard
-            self.__dict__ = hazard.__dict__
-        else:
-            self.__dict__ = Hazard.concat([self, hazard]).__dict__
+        for haz in haz_list:
+            haz._check_events()
+
+        # check type, unit, and attribute consistency among hazards
+        haz_types = {haz.tag.haz_type for haz in haz_list if haz.tag.haz_type != ''}
+        if len(haz_types) > 1:
+            raise ValueError(f"The given hazards are of different types: {haz_types}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        self.tag.haz_type = haz_types.pop()
+
+        haz_classes = {type(haz) for haz in haz_list}
+        if len(haz_classes) > 1:
+            raise TypeError(f"The given hazards are of different classes: {haz_classes}. "
+                            "The hazards are incompatible and cannot be concatenated.")
+
+        units = {haz.units for haz in haz_list if haz.units != ''}
+        if len(units) > 1:
+            raise ValueError(f"The hazards use different units: {units}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        elif len(units) == 0:
+            units = {''}
+        self.units = units.pop()
+
+        attributes = sorted(set.union(*[set(vars(haz).keys()) for haz in haz_list]))
+        for attr_name in attributes:
+            if not all(hasattr(haz, attr_name) for haz in haz_list_nonempty):
+                raise ValueError(f"Attribute {attr_name} is not shared by all hazards. "
+                                 "The hazards are incompatible and cannot be concatenated.")
+
+        # append all tags (to keep track of input files and descriptions)
+        for haz in haz_list:
+            if haz.tag is not self.tag:
+                self.tag.append(haz.tag)
+
+        # map individual centroids objects to union
+        centroids = Centroids.union(*[haz.centroids for haz in haz_list])
```

Comment:
Should it be `haz_list` or `haz_list_nonempty`? Not sure what is conceptually more sensible.

Reply:
As I said above, I only use haz_list_nonempty for cases where it obviously doesn't make a difference whether to include them or not.

In this case, it makes a difference: When the hazards have been generated in a batch-kind of fashion, some of the hazards might contain no events, but they might still cover a region that is not covered by the other hazards.

Reply:
I agree. My question was whether this is the desired functionality (e.g. that the centroids of the empty hazard are added). I am fine with it.